iothub/device/src/InternalClient.cs

```diff
@@ -1349,44 +1217,151 @@ internal void OnReportedStatePatchReceived(TwinCollection patch)
             _desiredPropertyUpdateCallback(patch, _twinPatchCallbackContext);
         }
 
-        private Task EnableMethodAsync(CancellationToken cancellationToken)
+        #region Device Specific API
```

Comment:
There are some methods which are used in both Device and Module clients, and some which are client specific. I have separated the client specific methods into regions.

Reply:
In the Java, SDK, inheritance is used to make this a bit cleaner. File upload APIs live in device client since module clients can't do it, and sendEventAsync lives in internal client since both module and device client can do it. It's a bigger ask, but should we move away from composition and into inheritance (i.e. device client extends internal client, not device client has an instance of internal client)?

A switch like that would make this layer significantly less confusing

Reply:
Not sure I'd use inheritance, but partial classes would give you the same effect. They bring their own evil, but that'd be closer to what we should have had: sub clients.

Reply:
I agree, that was our plan at one point, where we would have the device client/ module client inherit from internal client, but we haven't gotten around to actually doing that. Our InternalClient is already "internal", so we aren't exposing those APIs to begin with, so public API surface wise, it wouldn't have a big impact.
It being a bigger work-item though, I am not sure on the prioritization for it. 

To me this felt like a good workaround for the time being.
That being said, if we prefer to give it a complete overhaul, I am all up for it.

Reply:
> Our InternalClient is already "internal", so we aren't exposing those APIs to begin with, so public API surface wise, it wouldn't have a big impact.

So, this statement by me was actually incorrect. InternalClient is internal, but if we were to have DeviceClient inherit from it, we would need to make InternalClient public. I am not particularly inclined to do that, since it is entirely an implementation detail, and it shouldn't be accessible to users.